src/persist-client/src/stats.rs

```diff
@@ -0,0 +1,116 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Aggregate statistics about data stored in persist.
+
+use std::collections::BTreeMap;
+
+use mz_persist::indexed::columnar::ColumnarRecords;
+use mz_persist_types::columnar::{ColumnFormat, PartEncoder, Schema};
+use mz_persist_types::part::{Part, PartBuilder};
+use mz_persist_types::Codec;
+
+use crate::internal::encoding::Schemas;
+
+/// Logic for filtering parts in a read as uninteresting purely from stats about
+/// the data held in them.
+pub trait PartFilter<K, V>: 'static {
+    /// True if the part should be fetched, false if it should be skipped.
+    fn should_fetch(&mut self, stats: &PartStats) -> bool;
+}
+
+/// An implementation of [PartFilter] that always returns true.
+#[derive(Debug)]
+pub struct FetchAllPartFilter;
+
+impl<K, V> PartFilter<K, V> for FetchAllPartFilter {
+    fn should_fetch(&mut self, _stats: &PartStats) -> bool {
+        true
+    }
+}
+
+/// Aggregate statistics about data contained in a [Part].
+///
+/// TODO(mfp): PartialEq, Eq, PartialOrd, Ord are sus.
+#[derive(Debug, Default, PartialEq, Eq, PartialOrd, Ord)]
+pub struct PartStats {
+    // TODO(mfp): Only contains Option<u64> columns in this initial skeleton.
```

Comment:
I don't quite know how we extend this to other types, but I really like `PartStats` being a properly owned Persist object. when I prototyped reusing Parquet, it felt tricky to work with since it always required the user's Schema to make sense of its contents, which wasn't a dealbreaker, but it felt like we didn't "own" the contents in the same way